calicoctl/commands/migrate/import.go

```diff
@@ -0,0 +1,409 @@
+	}
+
+	// First chunk should be the v3 resource YAML.
+	// Second chunk should give the cluster info resource.
+	// Last chunk should be the IPAM JSON.
+	return split[0], split[1], split[2], nil
+}
+
+func checkCalicoResourcesNotExist(args map[string]interface{}, c client.Interface) error {
+	// Loop through all the v3 resources to see if anything is returned
+	extendedV3Resources := append(allV3Resources, "clusterinfo")
+	for _, r := range extendedV3Resources {
+		// Skip nodes since they are backed by the Kubernetes node resource
+		if r == "nodes" {
+			continue
+		}
+
+		// Create mocked args in order to retrieve Get resources.
+		mockArgs := map[string]interface{}{
+			"<KIND>":   r,
+			"<NAME>":   []string{},
+			"--config": args["--config"].(string),
+			"--export": false,
+			"--output": "ps",
+			"get":      true,
+		}
+
+		if _, ok := namespacedResources[r]; ok {
+			mockArgs["--all-namespaces"] = true
+		}
+
+		// Get resources
+		results := common.ExecuteConfigCommand(mockArgs, common.ActionGetOrList)
+
+		// Loop through the result lists and see if anything exists
+		for _, resource := range results.Resources {
+			if meta.LenList(resource) > 0 {
+				return fmt.Errorf("Found existing Calico %s resource", results.SingleKind)
+			}
+
+			if results.FileInvalid {
+				return fmt.Errorf("Failed to execute command: %v", results.Err)
+			} else if results.Err != nil {
+				return fmt.Errorf("Failed to retrieve %s resources during datastore check: %v", resourceDisplayMap[r], results.Err)
+			}
+		}
+	}
+
+	// Check if any IPAM resources exist
+	ipam := NewMigrateIPAM(c)
+	err := ipam.PullFromDatastore()
+	if err != nil {
+		return fmt.Errorf("Failed to retrieve IPAM resources during datastore check: %s", err)
+	}
+
+	if !ipam.IsEmpty() {
+		return fmt.Errorf("Found existing IPAM resources")
+	}
+
+	return nil
+}
+
+func updateClusterInfo(ctx context.Context, c client.Interface, clusterInfoJson []byte) error {
+	// Unmarshal the etcd cluster info resource.
+	migrated := apiv3.ClusterInformation{}
+	err := json.Unmarshal(clusterInfoJson, &migrated)
+	if err != nil {
+		return fmt.Errorf("Error reading exported cluster info for migration: %s", err)
+	}
+
+	// Get the "default" cluster info resource.
+	clusterinfo, err := c.ClusterInformation().Get(ctx, "default", options.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("Error retrieving current cluster info for migration: %s", err)
+	}
+
+	// Update the calico version and cluster GUID.
+	clusterinfo.Spec.ClusterGUID = migrated.Spec.ClusterGUID
+	clusterinfo.Spec.CalicoVersion = migrated.Spec.CalicoVersion
+	_, err = c.ClusterInformation().Update(ctx, clusterinfo, options.SetOptions{})
+	if err != nil {
+		return fmt.Errorf("Error updating current cluster info for migration: %s", err)
+	}
+
+	return nil
+}
+
+func updateV3Resources(cf string, data []byte) error {
+	// Create tempfile so the v3 resources can be created using Apply
```

Comment:
This is probably fine, but I was originally imagining we'd do this all through libcalico-go rather than writing to file and pushing it through all of the calicoctl processing.

Again, probably fine for now if it's working. Might find a reason later to do it differently though, e.g., if there is some calicoctl validation that we don't want in this codepath.

Reply:
Only reason I did it this way is that it let me reuse a lot of the code as it was written so I wouldn't need to duplicate code to handle things like partially applying some of the resources or parsing the YAML files into the appropriate resources before applying them.

Reply:
yeah, understood. thinks this is good for now.